qiskit_experiments/calibration/calibration_definitions.py

```diff
@@ -0,0 +1,382 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Class to store the results of a calibration experiments."""
+
+import copy
+from datetime import datetime
+import dataclasses
+from typing import Tuple, Union, List, Optional, Type
+import pandas as pd
+
+from qiskit.circuit import Gate
+from qiskit import QuantumCircuit
+from qiskit.pulse import Schedule, DriveChannel, ControlChannel, MeasureChannel
+from qiskit.pulse.channels import PulseChannel
+from qiskit.circuit import Parameter
+from .exceptions import CalibrationError
+from .parameter_value import ParameterValue
+
+
+class CalibrationsDefinition:
+    """
+    A class to manage schedules with calibrated parameter values.
+    Schedules are stored in a dict and are intended to be fully parameterized,
+    including the index of the channels. The parameters are therefore stored in
+    the schedules. The names of the parameters must be unique. The calibrated
+    values of the parameters are stored in a dictionary.
+    """
+
+    def __init__(self, backend):
+        """
+        Initialize the instructions from a given backend.
+
+        Args:
+            backend: The backend from which to get the configuration.
+        """
+
+        self._n_qubits = backend.configuration().num_qubits
+        self._n_uchannels = backend.configuration().n_uchannels
+        self._config = backend.configuration()
+        self._params = {'qubit_freq': {}}
+        self._schedules = {}
+        self.backend = backend
+
+        # Populate the qubit frequency estimates
+        now = datetime.now()
+        for qubit, freq in enumerate(backend.defaults().qubit_freq_est):
+            val = ParameterValue(freq, now)
+            self.add_parameter_value('qubit_freq', val, DriveChannel(qubit))
+
+    def schedules(self) -> pd.DataFrame:
+        """
+        Return the schedules in self in a data frame to help
+        users manage their schedules.
+
+        Returns:
+            data: A pandas data frame with all the schedules in it.
+        """
+        data = []
+        for name, schedule in self._schedules.items():
+            data.append({'name': name,
+                         'schedule': schedule,
+                         'parameters': schedule.parameters})
+
+        return pd.DataFrame(data)
```

Comment:
Why it generates dataframe on the fly rather than storing data as dataframe? I think dataframe is upper compatible of dictionary, i.e. dataframe provides filtering functionality and flexibility of data structure in addition to the connectivity to SQL database. So there is no drawback of using dataframe as internal data storage. I agree dict is simpler, but I don't prefer nested version.

If we need table view, it's also possible to show schedules in stdout with line feed.

Reply:
The current implementation makes it very transparent what is needed since parameter values are defined by the `ParameterValue` class. If we use a dataframe to store the data then the structure becomes less transparent as you can have almost anything as a column. By contrast `ParameterValue` makes it clear what the calibration module requires.